libavcodec/libaribcaption.c

```diff
@@ -0,0 +1,1166 @@
+ */
+
+#include "avcodec.h"
+#include "codec_internal.h"
+#include "internal.h"
+#include "libavcodec/ass.h"
+#include "libavutil/avstring.h"
+#include "libavutil/avutil.h"
+#include "libavutil/thread.h"
+#include "libavutil/log.h"
+#include "libavutil/opt.h"
+
+#include <aribcaption/aribcaption.h>
+
+#if !defined(DEFAULT_FONT_ASS)
+# define DEFAULT_FONT_ASS "sans-serif"
+#endif
+
+#define ARIBC_BPRINT_SIZE_INIT         64
+#define ARIBC_BPRINT_SIZE_MAX          (8 * 1024)
+#define ARIBC_ALPHA_MAX_NUM            4
+#define ARIBC_ALPHA_DEFAULT_FRONT      0xFF
+#define ARIBC_ALPHA_DEFAULT_BACK       0x80
+
+#define ARIBCC_COLOR_RGB(c)            ((c) & 0xFFFFFF)
+#define ARIBCC_COLOR_DIFF_RGB(c1,c2)   (((c1) ^ (c2)) & 0x00FFFFFF)
+#define ARIBCC_COLOR_DIFF_A(c1,c2)     (((c1) ^ (c2)) & 0xFF000000)
+
+#define CLUT_RGBA(r,g,b,a) (((unsigned)(a) << 24) | ((r) << 16) | ((g) << 8) | (b))
+#define CLUT_A(c)          (((c) >> 24) & 0xFF)
+#define CLUT_R(c)          (((c) >> 16) & 0xFF)
+#define CLUT_G(c)          (((c) >>  8) & 0xFF)
+#define CLUT_B(c)          ( (c)        & 0xFF)
+
+#define ARIBCC_COLOR_TO_CLUT_RGBA(c,a) (((ARIBCC_COLOR_A(c) ? ARIBCC_COLOR_A(c) : (a)) << 24) | \
+                                        (ARIBCC_COLOR_R(c) << 16) | \
+                                        (ARIBCC_COLOR_G(c) <<  8) | \
+                                        (ARIBCC_COLOR_B(c)))
+
+typedef struct ARIBCaptionContext {
+    AVClass *class;
+    AVCodecContext *avctx;
+    const AVPacket *avpkt;
+    AVSubtitle *sub;
+
+    aribcc_context_t *context;
+    aribcc_decoder_t *decoder;
+    aribcc_renderer_t *renderer;
+
+    enum AVSubtitleType subtitle_type;
+    enum aribcc_encoding_scheme_t encoding_scheme;
+    bool ass_single_rect;
+    char *font;
+    bool replace_fullwidth_ascii;
+    bool force_stroke_text;
+    bool ignore_background;
+    bool ignore_ruby;
+    float stroke_width;
+    bool replace_drcs;
+
+    int64_t pts;
+    AVRational time_base;
+    int canvas_width;
+    int canvas_height;
+    int plane_width;
+    int plane_height;
+    int frame_width;
+    int frame_height;
+    int bitmap_plane_width;
+    int bitmap_plane_height;
+    int font_size;
+    int charstyle;
+    int border_style;
+    int readorder;
+
+    aribcc_caption_t caption;
+    aribcc_render_result_t render_result;
+    uint32_t *clut;
+    int clut_idx;
+    int clut_overflow;
+    uint8_t clut_alpha[ARIBC_ALPHA_MAX_NUM];
+} ARIBCaptionContext;
+
+static void hex_dump_debug(void *ctx, const char *buf, int buf_size)
+{
+    int i;
+
+    for (i = 0; i < buf_size; i++) {
+        ff_dlog(ctx, "%02hhx ", buf[i]);
+        if (i % 16 == 15)
+            ff_dlog(ctx, "\n");
+    }
+    if (i % 16)
+        ff_dlog(ctx, "\n");
+}
+
+static void logcat_callback(aribcc_loglevel_t level, const char* message, void* userdata)
+{
+    ARIBCaptionContext *ctx = userdata;
+    int lvl;
+
+    if (ctx->decoder != NULL) {
```

Comment:
Any particular reason for this check?

Reply:
I don't remenber the details, but I caught some crashes without the check.
Probably, the callback might be called after the context has been freed.